Mathlib/Algebra/Module/Submodule/Pointwise.lean

```diff
@@ -277,4 +289,221 @@ scoped[Pointwise] attribute [instance] Submodule.pointwiseMulActionWithZero
+
+@[simp] lemma mem_empty_smul_submodule (x : M) : x ∈ (∅ : Set R) • N ↔ x = 0 := by
+  fconstructor
+  · intro hx
+    rw [mem_set_smul_submodule_def, Submodule.mem_sInf] at hx
+    exact hx ⊥ (λ r _ hr ↦ hr.elim)
+  · rintro rfl; exact Submodule.zero_mem _
+
+lemma singleton_set_smul_submodule_eq [SMulCommClass R R M] (r : R) :
+    ({r} : Set R) • N = (r • N : Submodule R M) := by
+  apply set_smul_submodule_eq_of_le
+  · rintro r m rfl hm; exact ⟨m, hm, rfl⟩
+  · rintro _ ⟨m, hm, rfl⟩
+    rw [mem_set_smul_submodule_def, Submodule.mem_sInf]
+    intro p hp; exact hp rfl hm
+
+lemma mem_singleton_set_smul_submodule [SMulCommClass R R M] (r : R) (x : M) :
+    x ∈ ({r} : Set R) • N ↔ ∃ (m : M), m ∈ N ∧ x = r • m := by
+  rw [singleton_set_smul_submodule_eq]
+  change x ∈ (r • N : Set M) ↔ _
+  rw [coe_pointwise_smul]
+  fconstructor <;>
+  · rintro ⟨m, hm, rfl⟩; exact ⟨m, hm, rfl⟩
+
+/-- subset of a ring has a multiplicative action on its submodules-/
+protected def pointwiseSetMulActionSubmodule [SMulCommClass R R M] :
+    MulAction (Set R) (Submodule R M) where
+  one_smul x := show {(1 : R)} • x = x from SetLike.ext fun m =>
+    (mem_singleton_set_smul_submodule _ _ _).trans ⟨by rintro ⟨_, h, rfl⟩; rwa [one_smul],
+      fun h => ⟨m, h, (one_smul _ _).symm⟩⟩
+  mul_smul s t x := le_antisymm
+    (set_smul_submodule_le _ _ _ <| by
+      rintro _ n ⟨a, b, ha, hb, rfl⟩ hn
+      rw [mul_smul]
+      apply mem_set_smul_submodule_of_mem_mem (mem1 := ha)
+      exact mem_set_smul_submodule_of_mem_mem (mem1 := hb) (mem2 := hn))
+    (set_smul_submodule_le _ _ _ <| by
+      rintro r m hr hm
+      have : SMulCommClass R R x := ⟨fun r s m => Subtype.ext <| show r • s • m.1 = s • r • m.1
+        from smul_comm _ _ _⟩
+      rw [mem_set_smul_submodule] at hm
+      obtain ⟨c, hc1, rfl⟩ := hm
+      delta Finsupp.sum
+      simp only [AddSubmonoid.coe_finset_sum, coe_toAddSubmonoid, SetLike.val_smul]
+      rw [Finset.smul_sum]
+      refine Submodule.sum_mem _ ?_
+      intro r' hr'
+      rw [← mul_smul]
+      apply mem_set_smul_submodule_of_mem_mem (mem2 := (c _).2)
+      exact ⟨r, r', hr, hc1 hr', rfl⟩)
+
+scoped[Pointwise] attribute [instance] Submodule.pointwiseSetMulActionSubmodule
+
+lemma set_smul_submodule_eq_iSup [SMulCommClass R R M] :
+    s • N = ⨆ (r ∈ s), r • N :=
+  set_smul_submodule_eq_of_le _ _ _
+    (fun r m hr hm => (show r • N ≤ _ by
+      rw [iSup_subtype', ← sSup_range]
+      exact le_sSup ⟨⟨r, hr⟩, rfl⟩) ⟨_, hm, rfl⟩) <| by
+    rw [iSup_subtype', ← sSup_range, sSup_le_iff]
+    rintro _ ⟨⟨x, hx⟩, rfl⟩ _ ⟨y, hy, rfl⟩
+    exact mem_set_smul_submodule_of_mem_mem (mem1 := hx) (mem2 := hy)
+
+/-- subset of a ring has a distributive multiplicative action on its submodules-/
+protected def pointwiseSetDistribMulActionSubmodule [SMulCommClass R R M] :
+    DistribMulAction (Set R) (Submodule R M) where
+  smul_zero s := show s • ⊥ = ⊥ from eq_bot_iff.mpr fun x hx => show x = 0 by
```

Comment:
I think `s • ⊥ = ⊥` should be separated out as a (simp) lemma since the simp normal form is `⊥` rather than `0`.

Reply:
added `set_smul_bot_submodule`